formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt

```diff
@@ -61,6 +63,20 @@ public fun JsonPrimitive(value: Number?): JsonPrimitive {
     return JsonLiteral(value, isString = false)
 }
 
+/**
+ * Creates [JsonPrimitive] from the given number.
+ *
+ * // TODO add more docs once functionality is stabilised.
+ */
+@ExperimentalSerializationApi
+public fun JsonLiteralNumber(value: String?): JsonPrimitive {
```

Comment:
> However, if string literal "null" is passed (i.e. value == "null"), function should throw IllegalArgumentException. This is motivated by the fact that when we serialize such a literal, if effective becomes indistinguishable from JsonNull and can't be deserialized back to JsonLiteral. We cannot know for sure if this was a user intent to do so or bug in validation. Better to prohibit such cases. IAE message can sound like this: "It is impossible to create a literal unquoted value of 'null'. If you want to create json null literal, use JsonNull object, otherwise, use quoted literal."

@sandwwraith Could you explain more about the above, please? I don't understand how a problem could occur.

I thought maybe it was because serialisation should be symmetric, so all `JsonPrimitive`s should decode to `JsonPrimitive`. But `JsonNull` is a subtype of `JsonPrimitive`, so that seems like it works to me.

Reply:
Hi!

Our concern is not about `null` and resulting `JsonNull`, but `"null"` *string* literal. 

The idea is the following -- with this PR, we enable bidirectional _unquoted_ string literal, meaning that we can both write them to `JsonElement` and parse them back.

The problem is simple -- in unquoted mode, both `null` and `"null"` have *the same* serialized form (`{"key": null}`) and cannot be distinguished when read back. It means that we have the following potential options:

* Allow both literals for writing (creating programmatically), but both of them as `JsonNull`. This may lead to an inconsistent data model, but it is consistent with `lenient` mode (https://github.com/Kotlin/kotlinx.serialization/issues/1600)
* Prohibit `"null"` literals as indistinguishable when reading it back. The possible workaround is to use quoted mode for such literals.

We'd like to stick with the second option, but we are open for discussion

Reply:
Sorry, I don't really follow. I'm asking so I can figure out how to write a unit test for this, because at the moment if I disable the exception

```kotlin
@ExperimentalSerializationApi
@Suppress("FunctionName")
public fun JsonUnquotedLiteral(value: String?): JsonPrimitive {
    return when (value) {
        null -> JsonNull
//        JsonNull.content -> throw JsonEncodingException("It is impossible to create a literal unquoted value of 'null'. If you want to create JSON null literal, use JsonNull object, otherwise, use JsonPrimitive")
        else -> JsonLiteral(value, isString = false, coerceToInlineType = jsonUnquotedLiteralDescriptor)
    }
}
```

none of the test fail, e.g. https://github.com/Kotlin/kotlinx.serialization/commit/d00bff2bbda69a17083480ec9daaaad5a54d2b42 (except for the test that is checking that `"null" throws an exception).

I think there are 3 options

1. `"null"` → `JsonNull`
2. `"null"` → `throw JsonEncodingException`
3. `"null"` → `JsonLiteral("null", isString = false)`

You say Option 1 can cause problems, Option 2 will prevent problems, but I think Option 3 is fine, and should work as expected.

Would 3 work?


```kotlin
@ExperimentalSerializationApi
@Suppress("FunctionName")
public fun JsonUnquotedLiteral(value: String?): JsonPrimitive {
    return when (value) {
        null -> JsonNull
        else -> JsonLiteral(value, isString = false, coerceToInlineType = jsonUnquotedLiteralDescriptor)
    }
}
```

Reply:
Here is the test that expected to "fail" (if one replaces `println` with actual asserts):

```
@Test
fun testThatNullAndNullStringAreDistinguishable() {
    val withNull = buildJsonObject { put("key", JsonNull) }
    val withNullString = buildJsonObject { put("key", JsonUnquotedLiteral("null")) }
    println(withNull == withNullString) // 'false' as expected

    val withNullSerilized = Json.encodeToString(withNull)
    val withNullStringSerialized = Json.encodeToString(withNullString)

    println(withNullSerilized == withNullStringSerialized) // Should be 'false', but is 'true

    val deserialized1 = Json.decodeFromString<JsonElement>(withNullSerilized)
    val deserialized2 = Json.decodeFromString<JsonElement>(withNullStringSerialized)
    println(deserialized1 == deserialized2) // Should be 'false', but is 'true'
}
```

We have two **semantically different** json objects, yet they are serialized to the same JSON string and, when parsed, are coerced into one json object, effectively losing information.

While looking harmless, debugging such an error in an enterprise-level software, especially where `JsonElement` is used indirectly when read (e.g. by polymorphic serializer or some man-in-the-middle that invokes `JsonDecoder.decodeJsonElement`)

Reply:
Thanks for the test! I've added a bit more logging to go over why I'm confused...

(I disabled the `value == "null" -> throw Exception` for these tests)

### JsonObject.toString()

```kotlin
val withNull = buildJsonObject { put("key", JsonNull) }
val withNullString = buildJsonObject { put("key", JsonUnquotedLiteral("null")) }
println(
    """
    withNull:                   $withNull
    withNullString:             $withNullString
    withNull == withNullString: ${withNull == withNullString}
    """.trimIndent()
)
```

```text
withNull:                   {"key":null}
withNullString:             {"key":null}
withNull == withNullString: false
```

Which seems wrong to me... They look the same. I think it makes sense that `JsonUnquotedLiteral("null") == JsonNull`.


### Json.encodeToString 

```kotlin
val withNullEncoded = Json.encodeToString(JsonElement.serializer(), withNull)
val withNullStringEncoded = Json.encodeToString(JsonElement.serializer(), withNullString)
println(
    """
    withNullEncoded:                          $withNullEncoded
    withNullStringEncoded:                    $withNullStringEncoded
    withNullEncoded == withNullStringEncoded: ${withNullEncoded == withNullStringEncoded}
    """.trimIndent()
)
```

```text
withNullEncoded:                          {"key":null}
withNullStringEncoded:                    {"key":null}
withNullEncoded == withNullStringEncoded: true
```

This makes sense to me, they look the same. I don't see why it should be false. I'd expect that `JsonUnquotedLiteral("null")` produces outputs `null`.

### Json.decodeToString 

```kotlin
val withNullDecoded = Json.decodeFromString(JsonElement.serializer(), withNullEncoded)
val withNullStringDecoded = Json.decodeFromString(JsonElement.serializer(), withNullStringEncoded)

println(
    """
    withNullDecoded:                          $withNullDecoded
    withNullStringDecoded:                    $withNullStringDecoded
    withNullDecoded == withNullStringDecoded: ${withNullDecoded == withNullStringDecoded}
    """.trimIndent()
)
```

```text
withNullDecoded:                          {"key":null}
withNullStringDecoded:                    {"key":null}
withNullDecoded == withNullStringDecoded: true
```

Again, this also makes sense. They're both the same JSON content.

---

I think it's just difficult for me to imagine how this situation could happen. `JsonUnquotedLiteral()` is akin to disabling type-checking using `dynamic` in Kotlin/JS, so putting a minor safety net in for `"null"` seems unusual. The horse has already left the stable, as they say. 

I imagine that debugging issues basically be asking "are you using `JsonUnquotedLiteral()`? You probably shouldn't!" But you're the maintainers, not me! So if an exception on `"null"` makes it easier for you, then it makes sense.

Anyway, to wrap it up, I'll happily admit that I don't need to understand this! And I really want this PR ready as soon as possible, so I'll try and adapt the test you posted, and update the KDoc on `JsonUnquotedLiteral()`.